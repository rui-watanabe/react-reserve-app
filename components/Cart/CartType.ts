import { UserModelType } from '../../models/UserModel/UserType';
import { CartProductType } from '../../pagesType/CartType';
import { Token } from 'react-stripe-checkout';

type handleCheckoutFuncType = (token: Token) => void;
type handleRemoveFromCart = (productId: string) => void;

export interface CartItemListProps {
  products: CartProductType;
  user: UserModelType;
  handleRemoveFromCart: handleRemoveFromCart;
  success: boolean;
}

export interface CartSummaryProps {
  products: CartProductType;
  handleCheckout: handleCheckoutFuncType;
  success: boolean;
}
