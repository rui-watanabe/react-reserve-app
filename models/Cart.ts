import mongoose from 'mongoose';
import { CartModelType } from './ModelsType';

const { ObjectId, Number } = mongoose.Schema.Types;

const CartSchema = new mongoose.Schema({
  user: {
    type: ObjectId,
    ref: 'User',
  },
  products: [
    {
      quantity: {
        type: Number,
        default: 1,
      },
      product: {
        type: ObjectId,
        ref: 'Product',
      },
    },
  ],
});

export default mongoose.models.Cart ||
  mongoose.model<CartModelType>('Cart', CartSchema);
