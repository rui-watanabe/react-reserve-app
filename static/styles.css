:root {
  --bodyBg: #f0f4f8;
  --menuBg: #008080;
  --progressBar: #54d1db;
}

body {
  background: var(--bodyBg);
}

#menu {
  background: var(--menuBg);
}

/* Overrides default styled accordion styles (too much bottom padding) */
.ui.styled.accordion .content {
  margin: 0;
  padding: 0.5em 1em 0.75em !important;
}

.ui.card > .image:not(.ui) > img, .ui.cards > .card > .image:not(.ui) > img  {
  height: 212px;
}
