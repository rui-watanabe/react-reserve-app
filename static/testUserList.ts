import { UserModelType } from '../models/Types/UserType';

export const testUserList: UserModelType[] = [
  // testUser is 1xxxxxxxxx group
  {
    _id: '20000000001',
    createdAt: new Date(),
    updatedAt: new Date(),
    name: 'test',
    email: 'test@gmail.com',
    password: '11111111',
    role: 'root',
  },
  {
    // testUser is 1xxxxxxxxx group
    _id: '20000000002',
    createdAt: new Date(),
    updatedAt: new Date(),
    name: 'test2',
    email: 'test2@gmail.com',
    password: '22222222',
    role: 'admin',
  },
  {
    // testUser is 1xxxxxxxxx group
    _id: '20000000003',
    createdAt: new Date(),
    updatedAt: new Date(),
    name: 'test',
    email: 'test@gmail.com',
    password: '33333333',
    role: 'user',
  },
];
