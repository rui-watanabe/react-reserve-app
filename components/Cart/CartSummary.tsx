import React from 'react';
import StripeCheckout from 'react-stripe-checkout';
import { Button, Segment, Divider, Message } from 'semantic-ui-react';
import calculateCartTotal from '../../utils/calculateCartTotal';
import { CartSummaryProps } from './CartType';

function CartSummary({ products, handleCheckout, success }: CartSummaryProps) {
  const [cartAmount, setCartAmount] = React.useState(0);
  const [stripeAmount, setStripeAmount] = React.useState(0);
  const [isCartEmpty, setCartEmpty] = React.useState(false);
  let stripeKey: string;

  React.useEffect(() => {
    const { cartTotal, stripeTotal } = calculateCartTotal(products);
    setCartAmount(cartTotal);
    setStripeAmount(stripeTotal);
    setCartEmpty(products.length === 0);
  }, [products]);

  if (typeof process.env.STRIPE_PUBLIC_KEY === 'string') {
    stripeKey = process.env.STRIPE_PUBLIC_KEY;
  } else {
    stripeKey = '';
  }

  return stripeKey != '' ? (
    <>
      <Divider />
      <Segment clearing size="large">
        <strong>Sub total:</strong> ${cartAmount}
        <StripeCheckout
          name="MernNextApp"
          amount={stripeAmount}
          image={products.length > 0 ? products[0].product.mediaUrl : ''}
          currency="USD"
          shippingAddress={true}
          billingAddress={true}
          zipCode={true}
          token={handleCheckout}
          stripeKey={stripeKey}
          triggerEvent="onClick"
        >
          <Button
            icon="cart"
            disabled={isCartEmpty || success}
            color="teal"
            floated="right"
            content="Checkout"
          />
        </StripeCheckout>
      </Segment>
    </>
  ) : (
    <Message negative>
      <Message.Header>We`re sorry we can`t apply that discount</Message.Header>
      <p>Setting ENV error!</p>
    </Message>
  );
}

export default CartSummary;
