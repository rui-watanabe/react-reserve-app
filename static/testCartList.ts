import { CartProductType } from '../src/pagesTypes/CartType';
import { testProductList } from './testProductList';

export const testCartList: CartProductType = [
  {
    quantity: 1,
    product: testProductList[0],
  },
  {
    quantity: 2,
    product: testProductList[1],
  },
];
