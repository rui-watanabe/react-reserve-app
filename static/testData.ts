import { ProductModelType } from '../models/Types/ProductType';
import { UserModelType } from '../models/Types/UserType';
import { CartProductType } from '../src/pagesTypes/CartType';
import { testProductList } from './testProductList';
import { testUserList } from './testUserList';

export const testProduct: ProductModelType = testProductList[0];

export const testTotalPages = Math.floor(testProductList.length / 9);

export const testUser: UserModelType = testUserList[2]; // 'user' role user

export const testCartProductList: CartProductType = [
  { product: { ...testProduct }, quantity: 1 },
  { product: { ...testProductList[1] }, quantity: 2 },
  { product: { ...testProductList[2] }, quantity: 3 },
];
