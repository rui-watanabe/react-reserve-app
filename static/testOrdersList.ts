import { ObjectId, ObjectID } from 'mongodb';
import { OrderModelType } from '../models/Types/OrderType';
import { testProductList } from './testProductList';
import { testUserList } from './testUserList';

export const testOrdersList: OrderModelType[] = [
  {
    _id: new ObjectId('200000000000000000000001'),
    createdAt: new Date(2020, 4, 1, 12, 30, 45),
    updatedAt: new Date(2020, 4, 1, 12, 30, 45),
    products: [
      {
        _id: new ObjectID('100000000000000000000001'),
        quantity: 1,
        product: testProductList[0],
      },
      {
        _id: new ObjectID('100000000000000000000002'),
        quantity: 2,
        product: testProductList[1],
      },
    ],
    user: testUserList[2]._id, // 'user' role user
    email: testUserList[2].email,
    total: testProductList[0].price + testProductList[1].price * 2,
  },
  {
    _id: new ObjectID('200000000000000000000002'),
    createdAt: new Date(2020, 4, 1, 12, 30, 45),
    updatedAt: new Date(2020, 4, 1, 12, 30, 45),
    products: [
      {
        _id: new ObjectID('100000000000000000000003'),
        quantity: 1,
        product: testProductList[2],
      },
      {
        _id: new ObjectID('100000000000000000000004'),
        quantity: 2,
        product: testProductList[3],
      },
    ],
    user: testUserList[2]._id, // 'user' role user
    email: testUserList[2].email,
    total: testProductList[2].price + testProductList[3].price * 2,
  },
  {
    _id: new ObjectID('200000000000000000000003'),
    createdAt: new Date(2020, 4, 1, 12, 30, 45),
    updatedAt: new Date(2020, 4, 1, 12, 30, 45),
    products: [
      {
        _id: new ObjectID('100000000000000000000005'),
        quantity: 1,
        product: testProductList[4],
      },
      {
        _id: new ObjectID('100000000000000000000006'),
        quantity: 2,
        product: testProductList[5],
      },
    ],
    user: testUserList[2]._id, // 'user' role user
    email: testUserList[2].email,
    total: testProductList[4].price + testProductList[5].price * 2,
  },
];
