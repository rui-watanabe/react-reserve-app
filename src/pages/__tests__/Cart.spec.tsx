import React from 'react';
import renderer from 'react-test-renderer';
import { testCartProductList, testUser } from '../../../static/testData';
import Cart from '../cart';

test('Cart', () => {
  const component = renderer.create(
    <Cart products={testCartProductList} user={testUser} />,
  );
  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});
