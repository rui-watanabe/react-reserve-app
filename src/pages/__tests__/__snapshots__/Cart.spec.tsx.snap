// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart 1`] = `
<div
  className="ui segment"
>
  <div
    className="ui divided items"
  >
    <div
      className="item"
      fluid="true"
    >
      <div
        className="image"
      >
        <img
          src="https://res-5.cloudinary.com/dwpujv6in/image/upload/c_pad,dpr_2.0,f_auto,h_930,q_auto,w_930/v1/media/catalog/product/f/d/fd1_lngchr_bh_frontlow-field-lounge-chair-tait-blush.jpg"
        />
      </div>
      <div
        className="content"
      >
        <a
          className="header"
          onClick={[Function]}
        >
          Bernie Gledhill
        </a>
        <div
          className="meta"
        >
          1 x $1399.02
        </div>
        <button
          className="ui basic icon right floated button"
          onClick={[Function]}
        >
          <i
            aria-hidden="true"
            className="remove icon"
            onClick={[Function]}
          />
        </button>
      </div>
    </div>
    <div
      className="item"
      fluid="true"
    >
      <div
        className="image"
      >
        <img
          src="https://res-3.cloudinary.com/dwpujv6in/image/upload/c_pad,dpr_2.0,f_auto,h_930,q_auto,w_930/v1/media/catalog/product/d/s/ds1_lngchr_ca-deep-thoughts-lounge-chair-camel-leather.jpg"
        />
      </div>
      <div
        className="content"
      >
        <a
          className="header"
          onClick={[Function]}
        >
          Gwendolen Bickerstaffe
        </a>
        <div
          className="meta"
        >
          2 x $1599.93
        </div>
        <button
          className="ui basic icon right floated button"
          onClick={[Function]}
        >
          <i
            aria-hidden="true"
            className="remove icon"
            onClick={[Function]}
          />
        </button>
      </div>
    </div>
    <div
      className="item"
      fluid="true"
    >
      <div
        className="image"
      >
        <img
          src="https://res-1.cloudinary.com/dwpujv6in/image/upload/c_pad,dpr_2.0,f_auto,h_930,q_auto,w_930/v1/media/catalog/product/r/u/ru1_rton69_m1_overhead-right-on-6-9-rug-color-mix-1.jpg"
        />
      </div>
      <div
        className="content"
      >
        <a
          className="header"
          onClick={[Function]}
        >
          Desi Avramovitz
        </a>
        <div
          className="meta"
        >
          3 x $699.57
        </div>
        <button
          className="ui basic icon right floated button"
          onClick={[Function]}
        >
          <i
            aria-hidden="true"
            className="remove icon"
            onClick={[Function]}
          />
        </button>
      </div>
    </div>
  </div>
  <div
    className="ui negative message"
  >
    <div
      className="header"
    >
      We\`re sorry we can\`t apply that discount
    </div>
    <p>
      ENV type error!
    </p>
  </div>
</div>
`;
